T3896(adjacent): Fix ocserv local user requirement, add groupconfig

sempervictus opened the pull request:
From ocserv documentation:
```
If the groupconfig option is set, then config-per-user will be
overriden, and all configuration will be read from radius. That
also includes the Acct-Interim-Interval, and Session-Timeout
values.
```
Implement yes/no configuration and parameter handling during jinja
rendering.

Fix bug wherein openconnect-server configuration requires creation
of local user accounts even when RADIUS authentication is used.

Testing:
  Set the groupconfig=yes param and observed change in generated
/run/ocserv/ocserv.conf.
  Removed the local users via `delete vpn openconnect
authentication local-users` and observed commit & service operation

<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T3896

## Component(s) name
ocserv

## Proposed changes
Bug fixes during code review to determine how we can implement per-group config and RADIUS-based group authorization:
1. Remove local openconnect account requirement when using RADIUS
2. Permit setting `groupconfig` parameter in `/run/ocserv/ocserv.conf` for the `radius` setting

## How to test
```
# delete vpn openconnect authentication local-users
# set vpn openconnect authentication radius groupconfig yes
# commit
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
-- Run during build of the package
- [ ] My commit headlines contain a valid Task id
-- This is tangential to the task, done while grokking the code in praparation
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

Commit:
T3896: Drop cserv local user req, add groupconfig

From ocserv documentation:
```
If the groupconfig option is set, then config-per-user will be
overriden, and all configuration will be read from radius. That
also includes the Acct-Interim-Interval, and Session-Timeout
values.
```
Implement yes/no configuration and parameter handling during jinja
rendering.

Fix bug wherein openconnect-server configuration requires creation
of local user accounts even when RADIUS authentication is used.

Testing:
  Set the groupconfig=yes param and observed change in generated
/run/ocserv/ocserv.conf.
  Removed the local users via `delete vpn openconnect
authentication local-users` and observed commit & service operation

Commit:
T3896: Use group selector and forced dns tunneling

Enterprise RADIUS configurations often utilize group selectors for
authentication and attribute distribution for connecting clients.

Ocserv implements this functionality via the `select-group` config
file attribute, repeating for multiple groups. When a user selects
their membership group and the request is passed to the RADIUS
server, ocserv will match the returned Class attribute against the
value selected by the user. This functionality also works for local
group membership resolution, although VyOS currently doesn't have
group membership configuration for this.

Expose the tunnel-all-dns option in the ocserv config file allowing
users who deploy default routes to select split-dns and those who
do not to enable full DNS tunneling.

Testing:
  Smoketests & build
  Configured groups in openconnect profile and verified existence
in /run/ocserv/ocserv.conf
  Configured forced dns tunneling and verified presence of setting
in /run/ocserv/ocserv.conf

sever-sever commented:
Delete please Boolean “yes|no”
It will be enough just
`set vpn openconnect authentication radius groupconfig`

sever-sever commented on data/templates/ocserv/ocserv_config.j2:
```diff
@@ -7,7 +7,11 @@ run-as-user = nobody
 run-as-group = daemon
 
 {% if "radius" in authentication.mode %}
+{%     if "yes" in authentication.radius.groupconfig %}
```
Yes, No  not required

sever-sever commented on interface-definitions/vpn-openconnect.xml.in:
```diff
@@ -144,6 +154,26 @@
+                        <list>yes no</list>
+                      </completionHelp>
+                      <valueHelp>
+                        <format>yes</format>
```
Yes,No. not required, we should avoid Boolean values

sever-sever commented:
Could you read https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md#submit-a-patch

Once there is an entry in Phabricator, you should reference its id in your commit message, as shown below:

```
ddclient: T1030: auto create runtime directories
Jenkins: add current Git commit ID to build description
```

If there is no [Phabricator](https://phabricator.vyos.net/) reference in the commits of your pull request, we have to ask you to amend the commit message. Otherwise we will have to reject it.

sempervictus commented:
@sever-sever  Sorry, referenced it in the PR name since the actual task was made by someone else.
Can rebase with updated name on the next push.

I'm not up to speed on the syntax and approach in the XML files, so need a bit more guidance on what you mean by avoiding boolean values (the underlying config uses `true`|`false`). Whats the right way to implement this change if yes and no are not required?

Commit:
T3896: update groupconfig syntax per PR1463

Address @sever-sever's suggestion to refactor how groupconfig is
defined, parsed, and set (with his proposed conditional string
appending Py-sugar). Use the disable-mobike refactor as template
for XML simplification.

Testing:
  None yet

sempervictus commented:
Lets see what the bots say about building this iteration - _think_ i got @sever-sever's suggestions in correctly, but would appreciate a review when folks have a sec.

sever-sever commented on interface-definitions/vpn-openconnect.xml.in:
```diff
@@ -50,6 +50,16 @@
                   </leafNode>
                 </children>
               </node>
+              <leafNode name="groups">
```
Could we rename it to `group`? Even if it has "multiple" values, it should be just "group"
For example, as we have it in the firewall, `set firewall group xxx`
We have only a couple of exceptions like `interfaces` and `protocols`

c-po replied:
Please use singular `group`

sempervictus replied:
Roger

sever-sever commented on interface-definitions/vpn-openconnect.xml.in:
```diff
@@ -144,6 +154,11 @@
                     </properties>
                     <defaultValue>2</defaultValue>
                   </leafNode>
+                  <leafNode name="groupconfig">
```
Maybe `group-config` will be feet better
Do we really need this extra option if we have the `group`? 
```
If the groupconfig option is set, then config-per-user will be overriden, and all configuration will be read from radius
```
Am I understanding correctly that you mean some kind of `authentication mode`
```
vyos@r14# set vpn openconnect authentication mode 
Possible completions:
   local                Use local username/password configuration (OTP supported)
   radius               Use RADIUS server for user autentication

```

sempervictus replied:
Yes sir, its a separate function.
You  can define local groups, or RADIUS group checks, but not pull attributes such as IP address _from_ RADIUS itself.
Its why we need to expand the CLI to include per-group configurations (such as the client subnet, routes, NS) so as to permit RADIUS-based user+group auth and device-based assignment of VPN attributes based on the group.

sever-sever commented on interface-definitions/vpn-openconnect.xml.in:
```diff
@@ -278,6 +293,26 @@
+                <properties>
+                  <help>If the tunnel-all-dns option is set to yes, tunnel all DNS queries via the VPN. This is the default when a default route is set.</help>
+                  <completionHelp>
+                    <list>yes no</list>
```
I mentioned about `yes|no` before, it is bad CLI syntax, so we should avoid it (`yes|no, True|False, enable|disable, 1|0`) where it is possible
@dmbaturin What do you think?

c-po replied:
Please use a `<valueLess/>` node. We stopped using booleans. The new style is: If not is present -> true, otherwise false

sempervictus replied:
Roger

sever-sever commented on src/conf_mode/vpn_openconnect.py:
```diff
@@ -57,15 +57,16 @@ def get_config():
-        # Not every configuration requires OTP settings
-        if ocserv['authentication']['local_users']['username'][user].get('otp'):
-            ocserv['authentication']['local_users']['username'][user]['otp'] = dict_merge(default_ocserv_usr_values, ocserv['authentication']['local_users']['username'][user]['otp'])
+    if "local" in ocserv["authentication"]["mode"]:
```
Just a small notice
Why do you use in one place single quotes and in another double quotes?

sempervictus replied:
Sloppiness mostly - if there's a preferred syntax, happy to try and use that. If we have a linter/formatter, even better.

sempervictus commented:
@c-po - the `tunnel-all-dns` option must be a boolean because its configuration changes dependent on how routes are assigned.
If the VPN service provides a default route, then it defaults to tunneling all DNS. If it does not, then it defaults to _not tunneling all DNS_ which is a problem on macs and such. Add to this the ability to assign routes _per-group/user_ via RADIUS or local configuration, and you end up requiring an explicit `on/off` switch for the behavior to enforce conformity across configurations.

c-po commented:
>  the tunnel-all-dns option must be a boolean because its configuration changes dependent on how routes are assigned.

As I see `tunnel-all-dns` is a boolean node with default = no, why does a `<valueLess/>` node not work in this case? To summarize what I understand is that `tunnel-all-dns = false` is different then not having `tunnel-all-dns` in the config at all?

sempervictus commented:
That is correct sir, it is different; and the difference depends on whether a default route has been issued to the client making an explicit setting needed to define behavior across (a set of or) all clients.
`As I see tunnel-all-dns is a boolean node with default = no` - this is only true when the client has not been issued a default route. So if the RADIUS server sends a framed route to `0.0.0.0/0` or we have one in the local user/group configs, then the default becomes `yes` for that specific at that time (RADIUS server may change login attribute return behavior depending on time of day or what-not).

c-po commented:
Can we combine it with the default route? And obly set it when the default route is present? so we then can control it as valueLess if its precondition is the default route

sempervictus commented:
We can have it handle default routes set by us (either as currently set in the top-level connection, or per-group/user as `ocserv` permits down the line), but would have to intercept RADIUS attributes to determine if any of the standard or vendor attributes for route distribution are passed back from the RADIUS host...

Commit:
T3896: update group syntax per PR1463
